src/cch/actor.rs

```diff
@@ -0,0 +1,818 @@
+            Contract::SimpleUDT,
+            hex::decode(
+                &self
+                    .config
+                    .wrapped_btc_type_script_args
+                    .trim_start_matches("0x"),
+            )
+            .map_err(|_| CchError::HexDecodingError)?
+            .as_ref(),
+        )
+        .into();
+        let mut order = SendBTCOrder {
+            expiry,
+            wrapped_btc_type_script,
+            fee_sats,
+            currency: send_btc.currency,
+            timestamp: duration_since_epoch.as_secs(),
+            ckb_final_tlc_expiry: self.config.ckb_final_tlc_expiry,
+            btc_pay_req: send_btc.btc_pay_req,
+            ckb_pay_req: Default::default(),
+            payment_hash: format!("0x{}", invoice.payment_hash().encode_hex::<String>()),
+            payment_preimage: None,
+            channel_id: None,
+            tlc_id: None,
+            amount_sats: amount_msat.div_ceil(1_000u128) + fee_sats,
+            status: CchOrderStatus::Pending,
+        };
+        order.generate_ckb_invoice()?;
+
+        state.orders_db.insert_send_btc_order(order.clone()).await?;
+        // TODO(now): save order and invoice into db: store.insert_invoice(invoice.clone())
+
+        Ok(order)
+    }
+
+    // On receiving new TLC, check whether it matches the SendBTC order
+    async fn handle_pending_received_tlc_notification(
+        &self,
+        state: &mut CchState,
+        tlc_notification: TlcNotification,
+    ) -> Result<()> {
+        let payment_hash = format!("{:#x}", tlc_notification.tlc.payment_hash);
+        log::debug!("[inbounding tlc] payment hash: {}", payment_hash);
+
+        let mut order = match state.orders_db.get_send_btc_order(&payment_hash).await {
+            Err(CchDbError::NotFound(_)) => return Ok(()),
+            Err(err) => return Err(err.into()),
+            Ok(order) => order,
+        };
+
+        if order.status != CchOrderStatus::Pending {
+            return Err(CchError::SendBTCOrderAlreadyPaid.into());
+        }
+
+        if tlc_notification.tlc.amount < order.amount_sats {
+            // TODO: split the payment into multiple parts
+            return Err(CchError::SendBTCReceivedAmountTooSmall.into());
+        }
+
+        order.channel_id = Some(tlc_notification.channel_id);
+        order.tlc_id = Some(tlc_notification.tlc.id.into());
+        state.orders_db.update_send_btc_order(order.clone()).await?;
+
+        let req = routerrpc::SendPaymentRequest {
+            payment_request: order.btc_pay_req.clone(),
+            timeout_seconds: BTC_PAYMENT_TIMEOUT_SECONDS,
+            ..Default::default()
+        };
+        log::debug!("[inbounding tlc] SendPaymentRequest: {:?}", req);
+
+        let mut client = state.lnd_connection.create_router_client().await?;
+        // TODO: set a fee
+        let mut stream = client.send_payment_v2(req).await?.into_inner();
```

Comment:
This will block the actor from processing other messages while issuing gRPC request. I think we can spawn a subtask and make this non-blocking. The problem is that after received gRPC response, we need to update the state. We can do that by add a message to the cch actor which can then process state update by handling this message.

Reply:
This function is aync. Awaiting on a inner future will not block the actor.

Reply:
Async function means that this code snippet will not block current thread, but this will still make the actor unable to process any other messages unrelated to this one, e.g. other SendBtc messages. This is because the function handle of this actor has still not returned, thus we have no chance to handle any other messages queued in the mailbox.